compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -805,14 +807,15 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
                             if ctor_vis.is_at_least(field_vis, &*self.r) {
                                 ctor_vis = field_vis;
                             }
+                            ret_fields.push(field_vis);
                         }
```

Comment:
```suggestion
                    let field_vis = self.resolve_visibility_speculative(&field.vis, true).unwrap_or(ty::Visibility::Public);
                    field_visibilities.push(field_vis);
                    if ctor_vis.is_at_least(field_vis, &*self.r) {
                        ctor_vis = field_vis;
                    }
```
The length of `field_visibilities` should ideally match the number of fields to be robust against other resolution errors, and here it's easy to make sure that they match.

Reply:
Both `push` and `if ctor_vis.is_at_least ...` can be moved to outside of the `resolve_visibility_speculative` condition, as the suggestion snippet shows.